Issue 5 - Birthdays

devmarcstorm opened the pull request:
Über die Navigation ist jetzt (für Admins) eine Geburtstagsliste verfügbar. 

Über dieses Template lassen sich jetzt (filterbar nach Gruppen) die Geburtstage aller Nutzer einsehen und auf dessen Nutzerprofile und Gruppen zugreifen.

Außerdem kann (über einen Link) eine ical generiert und heruntergeladen werden. 
(Nutzername, Gruppen, Geburtstage)

devmarcstorm committed:
Add BirthDate field and validate format

devmarcstorm committed:
Add translation for birthdate field

devmarcstorm committed:
DateField and DateStringConverter

devmarcstorm committed:
Change .gitignore

devmarcstorm committed:
Add birthday list with group filter

devmarcstorm committed:
Update translations

devmarcstorm committed:
Fixed DateToString issues

devmarcstorm committed:
Fix group filter

devmarcstorm committed:
Sort birthday list by months

devmarcstorm committed:
Show age after birthday

devmarcstorm committed:
Create ical file to download

devmarcstorm committed:
Add icalender to requirements.txt

devmarcstorm committed:
Add birthday field to newUser form

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+#import datetime
```
Der Import ist noch übrig.

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+from icalendar import Calendar
+
+def calculate_age(birthdate):
+    birthdate = birthdate.strftime("%Y-%m-%d")
```
Wofür ist der `strftime`-Aufruf, wenn direkt danach wieder `strptime` genutzt wird?

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+def calculate_age(birthdate):
+    birthdate = birthdate.strftime("%Y-%m-%d")
+
+    if birthdate != '':
```
Aus `strftime` kommt immer valider Output, das hier wird also immer wahr sein.

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+
+from icalendar import Calendar
+
+def calculate_age(birthdate):
```
Darf hier `None` übergeben werden? Es wäre am besten, kurz zu prüfen, und bei `birthdate is None` direkt nichts zurückzugeben.

rixx replied:
Da diese Funktion nur für User gebraucht wird, wäre es vermutlich am Besten, sie an die User-Klasse umzuziehen. Dort könnte sie sogar als property leben:

```python
class User:
    ...
    @property
    def age(self):
        return ...
```

Dann kann anderweitig direkt `user.age` genutzt werden.

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+    if birthdate != '':
+        birthdate = datetime.strptime(birthdate, "%Y-%m-%d")
+        today = date.today()
+        return (today.year - birthdate.year - ((today.month, today.day) < (birthdate.month, birthdate.day))) + 1
```
Haha, hacky.

Musste nicht ändern, aber so grundsätzlich: Es wäre vielleicht netter, den Boolean in eine eigene Variable auszulagern. So ist es doch nicht sooo gut lesbar, und lesbarer Code ist cooler als cleverer Code. ;)

rixx replied:
Darüber hinaus sieht es mir auch aus, als ob da nicht das richtige Ergebnis rauskommt (bzw vielleicht das Richtige für die Anzeige, aber wenn die Funktion "age" heißt, wäre da das wirkliche Alter das erwartete Ergebnis.)

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+    event = ''
+
+    for user in user_list:
+        summary = 'Birthday from {}'.format(user.name)
```
Internationalisiert, mit `_()`, wäre hier gut.

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+                groups += '{} '.format(group.name)
+
+        if date != '':
+            event += 'BEGIN:VEVENT\rSUMMARY:{}\rDTEND;VALUE=DATE:{}\rDTSTART;VALUE=DATE:{}\rUID:{}/27346262376@mxm.dk\rDESCRIPTION:{}\rPRIORITY:5\rEND:VEVENT\r'.format(summary, date, date, date, groups)
```
Hm, du schreibst jetzt hier ical von Hand, aber hast oben `Calendar` importiert. Welches solls sein?

rixx replied:
Wo kommt denn `UID:{}/27346262376@mxm.dk` her?

rixx replied:
Ich habe die ical-Spec auch überhaupt nicht im Kopf, aber das sieht mir aus, als würdest du einen Kalendereintrag am Geburtstagsdatum der Person anlegen, oder?

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -0,0 +1,38 @@
+    if event != '':
+        cal = 'BEGIN:VCALENDAR\r{}END:VCALENDAR'.format(event)
+        #f = open('ical/ode_birthdays.ics', 'wb')
+        #f.write(cal)
```
Die Kommentare sind noch übrig.

rixx commented on ode/blueprints/amu/forms.py:
```diff
@@ -35,6 +40,8 @@ class EditUserForm(Form):
 		userid = HiddenField(_('Username'), validators=[DataRequired()])
 		name = StringField(_('Full Name'), validators=[DataRequired()])
 
+		birthdate = DateField(_('Date of Birth'), format='%Y-%m-%d')
```
Hier vielleicht in die `description` das erwartete Format rein (oder zeigt das das `DateField` von sich aus an?)

rixx commented on ode/blueprints/amu/navigation.py:
```diff
@@ -23,6 +23,7 @@ def amu_navbar():
 				#View(_('New mailing list'), '.new_mailing_list'),
 				View(_('Aliases'), '.aliases'),
 				View(_('New Alias'), '.new_alias'),
+				View(_('Birthday Lists'), '.birthdays'),
```
Hm, da wäre fast die Frage, ob man das nur direkt auf der Userliste verlinkt. Gefühlt ist das nicht wichtig genug für einen eigenen Eintrag in der Navigation oben.

rixx commented on ode/blueprints/amu/templates/amu/birthdays.html:
```diff
@@ -0,0 +1,60 @@
+    <tr data-href="#">
+        <td><a href="{{url_for('.user',uid=user.userid)}}">{{user.name}}</a></td>
+        <td>
+            {% if user.birthdate.strftime("%Y-%m-%d") != '' %}
```
Da sollte eigentlich `{% if user.birthdate %}` reichen.

rixx commented on ode/blueprints/amu/templates/amu/birthdays.html:
```diff
@@ -0,0 +1,60 @@
+            <strong>{{_("Today is the birthday of this user")}}</strong>
+            {% endif %}
+            {% else %}
+            <strong>{{_("No birthday added")}}</strong>
```
Hier vielleicht einfach ein `-` anzeigen, das stört in der Liste weniger.

rixx commented on ode/blueprints/amu/templates/amu/birthdays.html:
```diff
@@ -0,0 +1,60 @@
+            {% if user_age_dict.get(date) != None%}
+            {{user_age_dict.get(date)}}
+            {% else %}
+            <strong>{{_("No birthday added")}}</strong>
```
Hier vielleicht einfach ein `-` anzeigen, das stört in der Liste weniger.

rixx commented on ode/model.py:
```diff
@@ -4,6 +4,32 @@
+	def value(self):
+		if len(self.__dict__['values']) == 1:
+			if self.values[0]:
+				return datetime.datetime.strptime(self.values[0], '%Y-%m-%d').date()
```
Hier fehlt noch ein `try`/`except` außen, um Fehler zu fangen, da ja hier beliebige Daten reinkommen können.

rixx commented on ode/model.py:
```diff
@@ -4,6 +4,32 @@
+	def __setattr__(self, key, value):
+		if key in ['value', '_init']:
+			if isinstance(value, datetime.date):
+				print(value)
```
Bitte das `print` noch raus.

devmarcstorm committed:
Cleaner Code and better uuid

- calculate_age is now a user property age
- uuids in the ical file are now unique
- small other changes

devmarcstorm committed:
Update translations

rixx commented on ode/blueprints/amu/birthday_functions.py:
```diff
@@ -1,38 +1,24 @@
             if user.dn in group.members:
                 groups += '{} '.format(group.name)
 
+        uuid = "{}-{}@{}".format(user.name, 'amu', socket.gethostname().lower())
```
Sicherheitshalber sollte auch das Geburtstatum mit rein, falls wir sonst später noch andere Termine generieren.

Außerdem bitte `user.name.replace(' ', '-')` – Leerzeichen sind zwar grundsätzlich in UIDs erlaubt, aber unüblich und unschön.

rixx commented on ode/blueprints/amu/views.py:
```diff
@@ -387,7 +387,8 @@ def birthdays():
-		age = calculate_age(ubd)
+		#age = calculate_age(ubd)
+		age = user.age
 		user_age_dict.update( {ubd.strftime("%Y-%m-%d") : age} )
```
Jetzt, wo du `user.age` hast, brauchst du das ganze `user_age_dict` nicht mehr.

rixx commented on ode/blueprints/amu/templates/amu/birthdays.html:
```diff
@@ -0,0 +1,60 @@
+            {% endif %}
+        </td>
+        <td>
+            {% if user_age_dict.get(date) != None%}
```
Hier kannst du jetzt einfach `` {{ user.age +1 }}`` nehmen.

rixx commented on ode/blueprints/amu/views.py:
```diff
@@ -374,3 +377,45 @@ def new_alias():
+	user_age_dict = {}
+	for user in user_list:
+		ubd = user.birthdate
+		#age = calculate_age(ubd)
```
Code comment 😉

devmarcstorm committed:
Fix small imperfections

rixx reviewed (approved):
Super! Ich kümmer mich bis zum Wochenende um den Merge-Konflikt und merge dann.

rixx commented:
Merged in 68d79e3
